From 21ac9566374c48aa84664165ae2eef02a7b2d7dd Mon Sep 17 00:00:00 2001
Message-Id: <21ac9566374c48aa84664165ae2eef02a7b2d7dd.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: =?UTF-8?q?Ville=20Syrj=C3=A4l=C3=A4?= <ville.syrjala at linux.intel.com>
Date: Fri, 20 Jun 2014 17:52:50 +0530
Subject: [PATCH 133/312] FROM_UPSTREAM [VPG]: drm/i915: Don't prefault the
 entire obj if the vma is smaller
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Take the minimum of the object size and the vma size and prefault
only that much. Avoids a SIGBUS when mmapping only a portion of the
object.

Prefaulting was introduced here:
 commit b90b91d87038f6b257b40a02b42ed4f9705e06f0
 Author: Chris Wilson <chris at chris-wilson.co.uk>
 Date:   Tue Jun 10 12:14:40 2014 +0100

    drm/i915: Prefault the entire object on first page fault

For: GMIN-1669
Upstream-repo: http://lists.freedesktop.org/archives/intel-gfx/2014-June/047450.html
Change-Id: Ic29903c8505621f1f97ea84a79fc9c61eea817e3
Cc: Chris Wilson <chris@chris-wilson.co.uk>
Signed-off-by: Ville Syrjälä <ville.syrjala@linux.intel.com>
---
 drivers/gpu/drm/i915/i915_gem.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index 4e72669..b6e4d97 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -1578,9 +1578,12 @@ int i915_gem_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
 	pfn >>= PAGE_SHIFT;
 
 	if (!obj->fault_mappable) {
+		unsigned long size = min_t(unsigned long,
+					   vma->vm_end - vma->vm_start,
+					   obj->base.size);
 		int i;
 
-		for (i = 0; i < obj->base.size >> PAGE_SHIFT; i++) {
+		for (i = 0; i < size >> PAGE_SHIFT; i++) {
 			ret = vm_insert_pfn(vma,
 				(unsigned long)vma->vm_start + i * PAGE_SIZE,
 				pfn + i);
-- 
1.7.9.5

